### ISSUE TITLE
Clickhouse Encoding Issue?

### ISSUE THREAD

**Issue by k----n** [woc-hack/tutorial#15]
I tried to use the following command similar to the tutorial: 
`clickhouse-client --host=da0 --query 'select sha1, author,comment from commits_all where time=1568656268 limit 1'`

Output:
```
?M??2ם\r???Ѽ?j;m?0/	狄狼 <910063@gmail.com>	0917\n
```

The sha1 output isn't a readable string?

**Comment by audrism** [comment]
sha1 needs to be read as a binary value (20bytes): to halve the amount of storage space it is stored as such. Using any sort of client (py/perl/etc) this can be easily achieved

**Comment by k----n** [comment]
Thanks.

For reference this is what I did on WoC to get it running in perl:
1. Install perl package manager to my home folder locally using: https://stackoverflow.com/a/2980715
2. Install the HTTP::Clickhouse package with: `cpanm HTTP::ClickHouse`

Run the following script as a test:
```
use HTTP::ClickHouse;

my $chdb =  HTTP::ClickHouse->new(
    host     => 'da0',
);


if ($chdb->select_array("select sha1, author,comment from commits_all where time=1568656268 limit 1")) {
    my $rows = $chdb->fetch_array();
    foreach my $row (@$rows) {
        for my $col (0 .. scalar(@$row) - 1) {
            if ($col == 0){
                my @sha1 = unpack('h*',@$row[$col]);
                print "@sha1\t";
            } else {
                print "@$row[$col]\t";
            }
        }
        print "\n";
    }
}
```

I didn't realize that the binary value was a hex string (low nybble first) until looking at the Python code.

Also note that the clickhouse client returns a reference to the array and not the array itself.

@audrism Is there a better way to install perl packages on WoC?

**Comment by k----n** [comment]
Actually it seems like I sometimes don't get the first 20 bytes when iterating through an array.

According to https://perlmaven.com/array-references-in-perl:
> If you are a C programmer, don't even think about pointer arithmetic. You can't do that in Perl.

@audrism I'm not really familiar with perl, do you have any idea how I can get the first 20 bytes of the array? Then take the rest of the bytes as another array?

**Comment by audrism** [comment]
Yes, now I remember more. CH uses leb128 encoding for binary data. 
Ws painful to import to it.


As for the output using RowBinary works well:
```
clickhouse-client --host=da1 --query 'select sha1 from commits_a limit 1 format RowBinary' | perl -ane 'chop(); $o=unpack "H*", $_; print "$o\n";'
```
I did not have much luck with the perl/py clients CH has distributed, used command-line one instead

**Comment by audrism** [comment]
Specifically in your case:
clickhouse-client --host=da1 --query 'select sha1, author from commits_all where time=1568656268 limit 1 format RowBinary' | \
perl -ane '$sha1=substr($_, 0, 20); $o=unpack "H*", $sha1;$rest=substr($_,20,length($_)-20); print "$o;$rest\n";'
```

**Comment by audrism** [comment]
Also, for more control, there is 
RowBinaryWithNamesAndTypes
https://clickhouse.tech/docs/en/interfaces/formats/#rowbinary

finally, if you look at /home/audris/lookup/importc2p.perl
there is an implementation of leb128 compatible with CH

**Comment by audrism** [comment]
btw, what kind of interface would make CH more usable?
On one hand, it is nice to have CH sql, on the other, pain to parse output. 

  -  One option is to bite the bullet and store sha1s in hex, 
  -  Another, to write/adapt existing clients (the provided ones go over http, so it is a bit hard to deal with binary data). 
  - Perhaps best use output RowBinaryWithNamesAndTypes to make the client-side parsing query-independent

**Comment by k----n** [comment]
Thanks, I think writing a client that deals with RowBinaryWithNamesAndTypes would be the best option.

I just tried the perl script and I guess it only works for single lines, e.g. `clickhouse-client --host=da1 --query "select sha1 from commits_all where match(comment, '^(CVE-(1999|2\d{3})-(0\d{2}[0-9]|[1-9]\d{3,}))$') limit 1 format RowBinary" | perl -ane 'chop(); $o=unpack "H*", $_; print "$o\n";'`

I'm not sure how to handle streaming inputs for when the query is like:
```
select sha1 
from commits_all 
where match(comment, '^(CVE-(1999|2\d{3})-(0\d{2}[0-9]|[1-9]\d{3,}))$')
format RowBinary
```

However, it is clear that clickhouse is a clear winner compared to the grepping the idx file (e.g. /data/All.blobs/commit_*.idx).

I am able to find CVE string matching commits like https://github.com/DanielRTeixeira/Powershell-Exploit/commit/501b9c9ea05506e6755dd173ec1b24cc4a83ffe8 way faster.

**Comment by k----n** [comment]
Update, I looked at the oscar.py code again and was able to come up with this query which seems to work quite well:
```
clickhouse-client --host=da1 --query "select lower(hex(sha1)), comment from commits_all where match(comment, '^(CVE-(1999|2\d{3})-(0\d{2}[0-9]|[1-9]\d{3,}))$')"
```

So I'm not sure if outputting to RowBinary is necessary?

P.S. I also do think importing the maps to clickhouse is beneficial in that it makes querying easier and faster.

**Comment by audrism** [comment]
RowBinary with no strings in it is "binary", so you need to read the entire output and covnert to hex, i.e., 
| perl -e '$s="";while(<STDIN>){$s.=$_;};$n=length($s)/20-1;for $i (0..$n){$sh=unpack H*, substr($s, $i*20, 20);print "$sh\n";}'

**Comment by audrism** [comment]
> So I'm not sure if outputting to RowBinary is necessary?
With RowBinary you have the exact bytes of the stored objects.  
 
> P.S. I also do think importing the maps to clickhouse is beneficial in that it makes querying easier and faster.

Do you have suggestions on which maps are first priority?

**Comment by k----n** [comment]
> RowBinary with no strings in it is "binary", so you need to read the entire output and covnert to hex, i.e.,
> | perl -e '$s="";while(){$s.=$_;};$n=length($s)/20-1;for $i (0..$n){$sh=unpack H*, substr($s, $i*20, 20);print "$sh\n";}'

Thanks!

> > So I'm not sure if outputting to RowBinary is necessary?
> > With RowBinary you have the exact bytes of the stored objects.
> 
> > P.S. I also do think importing the maps to clickhouse is beneficial in that it makes querying easier and faster.
> 
> Do you have suggestions on which maps are first priority?

I think combining the c2p and c2P into a single table would be interesting with p as an array column (array columns look promising: https://altinity.com/blog/harnessing-the-power-of-clickhouse-arrays-part-1).

I'm interested in seeing the performance of joining the commits_all table and the newly created c2p/c2P table when querying and if it's faster than calling the perl scripts for the same data.

**Comment by k----n** [comment]
It turns out clickhouse might not be that good at joins (e.g. https://stackoverflow.com/questions/40379948/how-make-join-table-in-clickhouse-db-faster), but it'd be interesting for p and P to be in the same table as all_commits.

Having all the "c" map info (e.g. c2b, c2p, c2P, c2cc, etc.) in one table seems like a good thing to try so that you don't have to do multiple lookups.
E.g. It seems like something like this https://github.com/woc-hack/tutorial#activity-7-suggested-by-the-audience in one query might be faster in clickup.

**Comment by audrism** [comment]
Ok, so I hear two requests: 

1) a client that parses an arbitrary CH query with  RowBinaryWithNamesAndTypes 

2) a more sophisticated c2pP table. CH needs to be revamped anyway.

Would you mind suggesting a schema for 2). Currently there is a commit table:
```
for h in da0 da1 da2 da3 da4; 
do echo "CREATE TABLE commits_q (sha1 FixedString(20), time Int32, tree FixedString(20), author String, parent String, comment String, content String) ENGINE = MergeTree() ORDER BY time" |clickhouse-client --host=$h
  echo "CREATE TABLE commits_all AS commits_q ENGINE = Distributed(da, default, commits_q, rand())" | clickhouse-client --host=$h
done
```
and c2p tables (pretty old):
```
for i in {0..31}; do clickhouse-client --max_partitions_per_insert_block=1000 --host=da3 --query "CREATE TABLE c2p_$i (date Date, sha1 FixedString(20), p String) ENGINE = MergeTree(date, sha1, 8192)"; done
```

and api import table:
```
for h in da0 da1 da2 da3 da4; 
do echo "CREATE TABLE b2cPtaPkgR (blob FixedString(20), commit FixedString(20), project String, time UInt32, author String, language String, deps String) ENGINE = MergeTree() ORDER BY time" |clickhouse-client --host=$h
  echo "CREATE TABLE b2cPtaPkgR_all AS b2cPtaPkgR ENGINE = Distributed(da, default, b2cPtaPkgR, rand())" | clickhouse-client --host=$h
done
```
